tests/test_transformer.py

```diff
@@ -0,0 +1,95 @@
+from __future__ import annotations
+
+import sys
+from enum import Enum
+from types import FunctionType  # noqa: F401
+from typing import Dict, Tuple, cast
+from unittest import skipIf
+
+from basedtyping import ForwardRef, Intersection
+from basedtyping.transformer import eval_type_based
+from typing_extensions import Annotated, Callable, Literal, TypeGuard, TypeIs, Union
+
+# ruff: noqa: PYI030, PYI030
+
+
+def validate(value: str, expected: object, *, string_literals=False):
+    assert (
+        eval_type_based(
+            ForwardRef(value),
+            globalns=cast(Dict[str, object], globals()),
+            string_literals=string_literals,
+        )
+        == expected
+    )
+
+
+@skipIf(sys.version_info <= (3, 10), "unsupported")  # type: ignore[no-any-expr]
+def test_literal():
+    validate("1 | 2", Union[Literal[1], Literal[2]])
+
+
+def test_literal_union():
+    validate("Union[1, 2]", Union[Literal[1], Literal[2]])
+
+
+def test_literal_literal():
+    validate("Literal[1]", Literal[1])
+
+
+def test_literal_str():
+    validate("'int'", int)
+    validate("Literal['int']", Literal["int"])
+    validate("'int'", Literal["int"], string_literals=True)
+    validate("Literal['int']", Literal["int"], string_literals=True)
+
+
+class E(Enum):
+    a = 1
+    b = 2
+
+
+@skipIf(sys.version_info <= (3, 10), "unsupported")  # type: ignore[no-any-expr]
+def test_literal_enum():
+    validate("E.a | E.b", Union[Literal[E.a], Literal[E.b]])
+
+
+def test_literal_enum_union():
+    validate("Union[E.a, E.b]", Union[Literal[E.a], Literal[E.b]])
+
+
+def test_tuple():
+    validate("(int, str)", Tuple[int, str])
+
+
+def test_typeguard():
+    validate("x is 1", TypeIs[Literal[1]])
+
+
+def test_typeguard_asymmetric():
+    validate("x is 1 if True else False", TypeGuard[Literal[1]])
+
+
+def test_callable():
+    validate("(str) -> int", Callable[[str], int])
+
+
+def test_function():
+    validate("def (str) -> int", Callable[[str], int])
+    validate("FunctionType[[str], int]", Callable[[str], int])
+
+
+def test_functiontype():
+    validate("FunctionType[[str], int]", Callable[[str], int])
+
+
+def test_intersection():
+    validate("int & str", Intersection[int, str])
+
+
+def test_nested():
+    validate("(1, 2)", Tuple[Literal[1], Literal[2]])
```

Comment:
what about:
- `(1, 2,)`
- `(1, 2,)`
- `(1, 2,),`
- `Foo[(1, 2,)]`
- `Foo[(1, 2)]`
- `Foo[(1, 2),]`
- `Foo[(1, 2,),]`
- `Foo[1, 2,]`

Reply:
`(1, 2,)` === `(1, 2)`

`Foo[(1, 2,)]` === `Foo[(1, 2)]` === `Foo[1, 2,]` === `Foo[1, 2]`

`Foo[(1, 2),]` === `Foo[(1, 2,),]`, i'll add a test for this one

Reply:
based, this found a bug with nested tuples